CIP-????/README.md

```diff
@@ -0,0 +1,511 @@
+---
+CIP: 145
+Title: Structured Contracts
+Category: Tools
+Status: Active
+Authors:
+  - Polina Vinogradova <polina.vinogradova@iohk.io>
+Implementors: []
+Discussions:
+  - https://github.com/cardano-foundation/CIPs/pull/963
+Created: 2024-04-15
+License: Apache-2.0
+---
+
+## Abstract
+
+At this time, there exists an [Agda model](https://github.com/IntersectMBO/formal-ledger-specifications) 
+of the Cardano ledger, which is used, among other things, as the source of truth 
+for the design of the Cardano ledger implementation, as well as for conformance testing. 
+This CIP describes a proposal for adding a framework for defining stateful computation called [structured contracts](https://omelkonian.github.io/data/publications/eutxo-struc.pdf) to this Agda specification model, 
+using a certain class of boolean Agda predicates to model ledger scripts. A user may instantiate 
+this framework by specifying the semantics of their desired stateful program, then defining Agda functions 
+corresponding to smart contracts required to implement these semantics. 
+Finally, to show the implementation is correct, fulfilling some proof obligations 
+concerting the definitions of the program semantics and smart contract functions. The purpose of this CIP is to define a principled approach to *correctly implementing a program on Cardano*, and equip users with the tools and the examples
+useful in developing formally verified Cardano applications. Any DApp developed on Cardano can be viewed 
+as an instance of our proposed framework, including anything from minting policies, to a basic (single-execution) 
+UTxO-locking script, to sophisticated stateful apps distributed across multiple UTxOs and scripts. We do not propose any 
+automation for defining structured contract instances, however, the scope of this CIP should include automated conversion 
+from completed Agda smart contract functions to Plutus code. 
```

Comment:
I'm skeptical that automated conversion can produce both readable and performant plutus code. It seems more feasible to use it as a reference implementation.

Reply:
Well, technically you could use agda2hs which should generate Haskell code that isn't different from hand-written code. In many cases that can probably be quite performant. But agda2hs has its own set of constraints on the Agda you can write, which gets in the way of having the proofs you want. So yeah, it's probably more realistic to have hand-written scripts.